src/postprocess/chi0.jl

```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
+# for the already-computed states, we add an explicit contribution
+# for the empty states, we solve a Sternheimer equation
+# (H-εn) δψn = - P_{ψ^⟂} δH ψn
+
+# The explicit term needs a careful consideration of stability.
+# Let <ψm|δψn> = αmn <δψm|δH|ψn>. αmn has to satisfy
+# fn αmn + fm αnm = ratio = (fn-fm)/(εn-εm)
+# The usual way is to impose orthogonality (=> αmn=-αnm),
+# but this means that αmn = 1/(εm-εn), which is unstable
+# Instead, we minimize αmn^2 + αnm^2 under the linear constraint above, which leads to
+# fn αmn = ratio * fn^2 / (fn^2 + fm^2)
+
+# This formula is very nice
+# - It gives a vanishing contribution fn αmn for empty states (note that α itself blows up, but it's compensated by fn)
+# - In the case where fn=1/0 or fm=0 we recover the same formulas as the ones with orthogonality
+# - It does not blow up for degenerate states
+function get_αmn(fm, fn, ratio)
+    ratio == 0 && return ratio
+    ratio * fn / (fn^2 + fm^2)
+end
+
+@views function solve_Ω(ham, ψ, occ, εF, eigenvalues, δHψ; kwargs_sternheimer=(cgtol=1e-6, verbose=false))
     basis  = ham.basis
+    model = basis.model
+    temperature = model.temperature
+    filled_occ = filled_occupation(model)
     T      = eltype(basis)
-    @assert basis.model.spin_polarization in (:none, :spinless, :collinear)
+    Nk = length(basis.kpoints)
 
-    n_spin = basis.model.n_spin_components
-    @assert 1 ≤ n_spin ≤ 2
+    # First compute δocc, δεF
+    δεF = zero(T)
+    δocc = [zero(occ[ik]) for ik = 1:Nk]
+    if temperature > 0
+        # First compute change in occupation without self-consistent Fermi δεF
+        D = zero(T)
+        for ik = 1:Nk
+            for (n, εnk) in enumerate(eigenvalues[ik])
+                enred = (εnk - εF) / temperature
+                δεnk = real(dot(ψ[ik][:, n], δHψ[ik][:, n]))
+                fpnk = (filled_occ 
+                        * Smearing.occupation_derivative(model.smearing, enred)
+                        / temperature)
+                δocc[ik][n] = δεnk * fpnk
+                D += fpnk * basis.kweights[ik]
+            end
+        end
+        D = mpi_sum(D, basis.comm_kpts) # equal to minus the total DOS
+        dos = sum(compute_dos(εF, basis, eigenvalues))
+        δocc_tot = mpi_sum(sum(basis.kweights .* sum.(δocc)), basis.comm_kpts)
+        δεF = δocc_tot / D
 
+        # Now add δεF contribution
+        for ik = 1:Nk
+            for (n, εnk) in enumerate(eigenvalues[ik])
+                enred = (εnk - εF) / temperature
+                δεnk = real(dot(ψ[ik][:, n], δHψ[ik][:, n]))
+                fpnk = (filled_occ 
+                        * Smearing.occupation_derivative(model.smearing, enred)
+                        / temperature)
+                δocc[ik][n] = (δεnk-δεF) * fpnk
+            end
+        end
+    end
+    # δεF ensures charge neutrality
+    @assert abs(mpi_sum(sum(sum, δocc), basis.comm_kpts)) < sqrt(eps(T))
+
+    # compute δψnk band per band
+    δψ = zero.(ψ)
+    for ik = 1:Nk
```

Comment:
`enumerate` or `map` ?

Reply:
not really because it does two things: increment D and assign docc